internal/util/fetch/fetch.go

```diff
@@ -0,0 +1,284 @@
+	"github.com/GoogleContainerTools/kpt/pkg/kptfile/kptfileutil"
+	"sigs.k8s.io/kustomize/kyaml/copyutil"
+	"sigs.k8s.io/kustomize/kyaml/errors"
+)
+
+// Command fetches a package from a git repository and copies it to a local
+// directory.
+type Command struct {
+	RepoSpec *git.RepoSpec
+
+	Destination string
+
+	Name string
+
+	Clean bool
+}
+
+// Run runs the Command.
+func (c Command) Run() error {
+	if err := (&c).DefaultValues(); err != nil {
+		return err
+	}
+
+	if _, err := os.Stat(c.Destination); !c.Clean && !os.IsNotExist(err) {
+		return errors.Errorf("destination directory %s already exists", c.Destination)
+	}
+
+	// normalize path to a filepath
+	if !strings.HasSuffix(c.RepoSpec.Dir, "file://") {
+		c.RepoSpec.Dir = filepath.Join(path.Split(c.RepoSpec.Dir))
+	}
+
+	err := cloneAndCopy(c.RepoSpec, c.Destination, c.Name, c.Clean)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// cloneAndCopy fetches the provided repo and copies the content into the
+// directory specified by dest. The provided name is set as `metadata.name`
+// of the Kptfile of the package.
+func cloneAndCopy(r *git.RepoSpec, dest, name string, clean bool) error {
+	if err := ClonerUsingGitExec(r); err != nil {
+		return errors.Errorf("failed to clone git repo: %v", err)
+	}
+	defer os.RemoveAll(r.Dir)
+
+	// delete the existing package if it exists
+	if clean {
+		if err := os.RemoveAll(dest); err != nil {
+			return errors.Wrap(err)
+		}
+	}
+
+	if err := copyutil.CopyDir(r.AbsPath(), dest); err != nil {
+		return errors.WrapPrefixf(err, "missing subdirectory %s in repo %s at ref %s\n",
+			r.Path, r.OrgRepo, r.Ref)
+	}
+
+	if err := upsertKptfile(dest, name, r); err != nil {
+		return errors.Wrap(err)
+	}
+	return nil
+}
+
+// ClonerUsingGitExec uses a local git install, as opposed
+// to say, some remote API, to obtain a local clone of
+// a remote repo.
+func ClonerUsingGitExec(repoSpec *git.RepoSpec) error {
+	// look for a tag with the directory as a prefix for versioning
+	// subdirectories independently
+	originalRef := repoSpec.Ref
+	if repoSpec.Path != "" && !strings.Contains(repoSpec.Ref, "refs") {
+		// join the directory with the Ref (stripping the preceding '/' if it exists)
+		repoSpec.Ref = path.Join(strings.TrimLeft(repoSpec.Path, "/"), repoSpec.Ref)
+	}
+
+	defaultRef, err := gitutil.DefaultRef(repoSpec.OrgRepo)
+	if err != nil {
+		return err
+	}
+
+	// clone the repo to a tmp directory.
+	// delete the tmp directory later.
+	err = clonerUsingGitExec(repoSpec)
+	if err != nil && originalRef != repoSpec.Ref {
+		repoSpec.Ref = originalRef
+		err = clonerUsingGitExec(repoSpec)
+	}
+
+	if err != nil {
+		if strings.HasPrefix(repoSpec.Path, "blob/") {
+			return errors.Errorf("failed to clone git repo containing /blob/, "+
+				"you may need to remove /blob/%s from the url:\n%v", defaultRef, err)
+		}
+		return errors.Errorf("failed to clone git repo: %v", err)
+	}
+
+	return nil
+}
+
+func clonerUsingGitExec(repoSpec *git.RepoSpec) error {
```

Comment:
How is it different from the public method above ? Comment would be helpful.

Reply:
Added some additional comments on this function and the public function.